### ISSUE TITLE
[data grid] Need help with customizing the DataGrid with `slots` prop

### ISSUE THREAD

**Issue by dizzyjaguar** [mui/mui-x#8194]
### Order ID or Support key 💳 (optional)

56801

### The problem in depth 🔍

I need a little guidance in customizing some `slots` on the Datagrid Pro. Mainly on how to provide the correct `clickHandlers` and `state` to the components that I'm replacing using the `slots` prop on my `DataGridPro`. For example on the `DataGrid-panel` that controls the `show/hide` of the columns, I want to replace the `baseSwitch` with a `Checkbox`. 

- I then use `baseSwitch: Checkbox,` or `baseSwitch: () => <Checkbox />` , but I'm not understanding the best way to provide the `state` and functionality from the `baseSwitch` to the `Checkbox. 
- Currently I'm looking through the code at how [GridColumnsPanel](https://github.com/mui/mui-x/blob/master/packages/grid/x-data-grid/src/components/panel/GridColumnsPanel.tsx) implements this.
- But this seems very inefficient on my part 😢 

I'm curious if there is an easier way to achieve this using `slots` or if I somehow missed something in the documentation that covers this. I have quite a few more customizations I need to do to various `slots` on this project so any guidance would be greatly appreciated 😄 

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  I'm using Chrome
  &&
  
System:
    OS: macOS 13.2.1
  Binaries:
    Node: 16.14.0 - ~/.nvm/versions/node/v16.14.0/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 9.1.1 - ~/.nvm/versions/node/v16.14.0/bin/npm
  Browsers:
    Chrome: 111.0.5563.64
    Edge: Not Found
    Firefox: 107.0
    Safari: 16.3
  npmPackages:
    @emotion/react: ^11.10.5 => 11.10.5
    @emotion/styled: ^11.10.5 => 11.10.5
    @mui/base:  5.0.0-alpha.111
    @mui/core-downloads-tracker:  5.11.1
    @mui/icons-material: ^5.11.0 => 5.11.0
    @mui/material: ^5.11.1 => 5.11.1
    @mui/private-theming:  5.11.1
    @mui/styled-engine:  5.11.0
    @mui/system:  5.11.1
    @mui/types:  7.2.3
    @mui/utils:  5.11.1
    @mui/x-data-grid:  6.0.0
    @mui/x-data-grid-pro: ^6.0.0 => 6.0.0
    @mui/x-license-pro:  6.0.0
    @types/react: ^18.0.26 => 18.0.26
    react: ^18.2.0 => 18.2.0
    react-dom: ^18.2.0 => 18.2.0
    typescript: ^4.9.3 => 4.9.4

```
</details>

**Comment by DanailH** [comment]
Hi @dizzyjaguar, did you try using the `slotProps` to pass in props to your custom components? You can check examples of how to use it here -> https://mui.com/x/react-data-grid/components/#overriding-components

**Comment by dizzyjaguar** [comment]
Hey @DanailH , I did explore that. Ended up figuring out what I needed so that I could easily replace components and still use the same handlers/state. 
What I was missing on my custom component was including the `props` that originally existed on the default component and spreading them into my custom component that is replacing that `slot`
```
function DataGridCheckbox(props: SwitchProps) {
  return <Checkbox color="marine" {...props} />
}
```

I think this could be super helpful to be explained clear in the [data grid component docs](https://mui.com/x/react-data-grid/components/)

I'd love to help contribute to this 😄 
Could you point me to contribution guidelines for this sort of thing.
Thanks so much

**Comment by DanailH** [comment]
Community contributions are always welcomed. You can check the guide here -> https://github.com/mui/mui-x/blob/master/CONTRIBUTING.md

**Comment by RomanFromEarth** [comment]
@DanailH I have to agree with @dizzyjaguar . The documentation for this is really not developed. I also have the problem that I want to replace the toggle in the ColumPanel with a checkbox. If possible, I would like to have all props and functionality automatically transferred to the component. Is something planned here? Or am I making mistakes in the application myself?

**Comment by michelengelen** [comment]
The question was solved in this thread: a component passed to a slot receives the props of the component it replaces, so spreading them is the whole trick — `function DataGridCheckbox(props: SwitchProps) { return <Checkbox {...props} /> }`. @RomanFromEarth the same applies to your switch-to-checkbox case: accept the incoming props and spread them into your component, and the state and handlers transfer automatically. The rebuilt [components docs](https://mui.com/x/react-data-grid/components/) now show this pattern, and `slotProps` plus [module augmentation](https://mui.com/x/react-data-grid/components/#custom-slot-props-with-typescript) cover passing extra props with correct types.

I will close this issue as answered. The documentation improvement it inspired is tracked in #8226 — please add feedback there if the current docs still leave the base slots unclear.
